sys/mips/82371AB.c

```diff
@@ -0,0 +1,141 @@
+/* 82371AB PCI-ISA bridge driver */
+
+#define KL_LOG KL_DEV
+#include <sys/mimiker.h>
+#include <sys/device.h>
+#include <sys/rman.h>
+#include <sys/bus.h>
+#include <mips/malta.h>
+#include <sys/pci.h>
+#include <sys/interrupt.h>
+#include <dev/piixreg.h>
+#include <dev/isareg.h>
+#include <sys/libkern.h>
+#include <sys/devclass.h>
+
+typedef struct intel_isa_state {
+  resource_t *io;
+  rman_t io_rman;
+} intel_isa_state_t;
+
+static resource_t *intel_isa_alloc_resource(device_t *dev, res_type_t type,
+                                            int rid, rman_addr_t start,
+                                            rman_addr_t end, size_t size,
+                                            rman_flags_t flags) {
+  assert(dev->bus == DEV_BUS_ISA);
+
+  if (type != RT_IOPORTS) {
+    device_t *idev = BUS_METHOD_PROVIDER(dev->parent, alloc_resource);
+    return BUS_METHODS(idev->parent)
+      .alloc_resource(idev, type, rid, start, end, size, flags);
+  }
+
+  assert(start >= IO_ICUSIZE);
+
+  rman_t *rman = NULL;
+  intel_isa_state_t *isa = dev->parent->state;
+  rman = &isa->io_rman;
+  bus_space_handle_t bh = isa->io->r_bus_handle;
+
+  resource_t *r =
+    rman_reserve_resource(rman, type, rid, start, end, size, 0, flags);
+
+  if (r == NULL)
+    return NULL;
+
+  /* We shouldn't do that for IRQs, but rn we'll never end up here when
+   * allocating them, because they get dispatched. */
+  r->r_bus_tag = generic_bus_space;
+  r->r_bus_handle = bh + resource_start(r);
+
+  /* IOPorts are always active */
+  if (bus_activate_resource(dev, r)) {
+    resource_release(r);
+    return NULL;
+  }
+
+  return r;
+}
+
+static int intel_isa_activate_resource(device_t *dev, resource_t *r) {
+  /* Neither IRQs in general or IOports on ISA require activation */
```

Comment:
Unfortunately, it won't handle all the scenarios correctly. Again, what if an ISA device allocates an inactive memory and wants to activate it later? Your solution is simply inconsistent. If the device allocates memory with the `RF_ACTIVE` flag, then you delegate the call to the parent which calls **its** activation function (which maps the resource). But if the device makes the same call without the mentioned flag and activates it later, then you call your **empty** activation function. As I said, I see two options:
1. If you don't want to remove the function, the you have to to the following
```C
{
  if (dev->r_type != RT_IOPORTS) {
    device_t *isab = dev->parent;
    return BUS_METHODS(isab->parent).activate_resource(isab, r);
  }
}
```
(As I mentioned in some previous comment, here we must do the explicit unwinding (not `bus_activate_resource`))
2. Just delegate the call to the parent.

Reply:
ISA devices are not supposed to allocate memory under current circumstances and allowing them to do so was not a scope of this driver, so I didn't bother with writing what would be a dead code.

In fact, what you propose is not just unnecessary - it's an inherently flawed solution. Even if we were to introduce memory allocations for ISA devices, no dispatching should be done. All procedures for managing memory resources should be handled locally within ISA driver, just like IO ports are. Dispatching activation procedures to another bus risks messing up a configuration for a device that's directly underneath it. To be precise - we don't want gt64120 driver to mess with any BARs when activating memory for ISA devices (Actually I'm pretty sure it would just panic on one of the asserts inside pci driver, because ISA bus is not a PCI device and doesn't have any BARs).

Reply:
If we don't allow ISA devices to allocate memory resources, then please add `assert(r->r_type != RT_MEMORY)` in activation/deactivation/release and `assert(type != RT_MEMORY)` in allocation.

Reply:
Here you go. I added some asserts. I do believe tho that adding them to activation/deactivation procedures was an overkill as allocating memory from ISA bus is impossible anyway.